### PR TITLE
[WIP] Add taints package

### PR DESCRIPTION
**Pull request by Tzif-Morgen**
#### Why we need this PR

Both FAR and SNR have duplicated taint-related code, particularly:                                                                              
  - Out-of-service taint detection and version checking logic
  - Generic taint manipulation utilities                                                                                                                                                                       
  - K8s version parsing for taint feature support
                                                                                                                                                                                                               
  Moving this code to common allows both projects to share tested, consistent implementations.                    
       
#### Changes made
1. Add `pkg/taints` package                                                                                                                                                                              
  New package providing shared taint utilities:
  - `TaintExists()` - Check if a taint exists in a taint slice                                                                                                                                                 
  - `FilterOutTaint()` - Return new slice without matching taints
  - `CreateOutOfServiceTaint()` - Create K8s out-of-service taint                                                                                                                                              
  - `InitOutOfServiceTaintFlagsWithRetry()` - Initialize version detection flags with retry logic
  - Package-level flags: `IsOutOfServiceTaintSupported` (K8s 1.26+), `IsOutOfServiceTaintGA` (K8s 1.28+)                                                                                                       
                                                                                                                                                                                                               
  Tests include version parsing with various formats and edge cases (11 test cases).                                                                                                                           
                                                                                                                                                                                                               
  2. Upgrade k8s dependencies to v0.33.7                                                                                                                                        
  Aligns common with FAR and SNR k8s versions, enables modern APIs like `wait.PollUntilContextTimeout`.                                                                                                        
                                                                                                                                                                                                               
  3. Fix lease test compatibility                                                                                                                                                                          
  Removed `Kind`/`APIVersion` assertions from lease tests. The new fake client (k8s v0.33) doesn't auto-populate TypeMeta fields like the old version did. These fields aren't used by manager logic. See
  [kubernetes-sigs/controller-runtime#3229](https://github.com/kubernetes-sigs/controller-runtime/pull/3229).                                                                                                  
   
  4. Rename `DeleteTaint` to `FilterOutTaint`                                                                                                                                                              
  Clearer name reflecting that it's a pure function returning a filtered slice, not modifying the original or deleting from a node. Each operator can use this function based on their specific needs.
                    

#### Which issue(s) this PR fixes

needed for  [RHWA-690](https://redhat.atlassian.net/browse/RHWA-690)



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added support for out-of-service taint management with automatic Kubernetes version detection.

* **Chores**
  * Updated Go runtime from version 1.20 to 1.24.0.
  * Upgraded all major dependencies including Kubernetes client libraries and controller-runtime to latest compatible versions.

* **Tests**
  * Expanded taint operation test coverage.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

**Comment by Tzif-Morgen** [comment]
we can also use PollWithContext instead of PollUntilContextTimeout to avoid all the k8s upgrades

**Review by razo7 — CHANGES_REQUESTED** [review]
## Code Review: Add taints package

### Must fix before merge

**1. Concurrency hazard — mutable package-level state**
`IsOutOfServiceTaintSupported` and `IsOutOfServiceTaintGA` are exported mutable package-level vars written by `InitOutOfServiceTaintFlagsWithRetry` and read by any consumer, with no synchronization. If two controllers in the same process call this, or a reader checks the flag while the retry loop is running, there's a data race.

Use `sync.Once` or `atomic.Bool`:
```go
type OutOfServiceTaintInfo struct {
    Supported bool
    GA        bool
}
```
Or at minimum protect with `sync.Once` since the values are write-once.

**2. Missing test coverage**
`TaintExists`, `FilterOutTaint`, and `CreateOutOfServiceTaint` are all untested. Since this is a shared library replacing duplicated code from FAR and SNR, comprehensive coverage is essential. Both FAR and SNR presumably had tests for their own implementations — those should migrate here.

**3. Nested if/else with shadowed `err` in `initOutOfServiceTaintFlags`**
The deep `if/else if/else` chain re-declares `err` in each `if` init statement, shadowing the outer variable. While the current code happens to work, it's fragile. Refactor to early returns.

### Should fix

**4. Discarded `PollUntilContextTimeout` error**
```go
_ = wait.PollUntilContextTimeout(ctx, interval, timeout, true, func(ctx context.Context) (bool, error) {
```
Intentional but undocumented. If the parent context is cancelled, the poll returns `ctx.Err()`, but you return the last internal `err`. Add a comment explaining the intent.

**5. Client-side `TimeAdded` in `CreateOutOfServiceTaint`**
Setting `TimeAdded` at Go function call time, not when the taint is applied to the node. In Kubernetes, `TimeAdded` is typically set by the API server. Consider removing it and letting the API server handle it.

**6. Stale comment**
`InitOutOfServiceTaintFlagsWithRetry` comment says "retry for a limited number of times" but the implementation is time-bound (2s interval, 10s timeout), not count-bound. Update the comment.

**7. Inconsistent `MatchTaint` call direction**
`TaintExists` calls `taint.MatchTaint(taintToFind)` while `FilterOutTaint` calls `taintToDelete.MatchTaint(&taints[i])`. `MatchTaint` is symmetric on Key+Effect so it works, but the inconsistency hurts readability. Pick one direction.

### Notes

- Tests use stdlib `testing` while the rest of the repo (`pkg/etcd`, `pkg/lease`) uses Ginkgo/Gomega — worth aligning for consistency.
- The dense test assertion in `setOutOfTaintFlags` checks three conditions in one `if` statement. Breaking into separate assertions would make failures easier to diagnose.
- No `suite_test.go` file, unlike other packages in the repo.

**Comment by Tzif-Morgen** [comment]
> 1. Concurrency hazard — mutable package-level state

Both FAR and SNR call it once in main.go at startup; Single-threaded initialization - no concurrent calls (and, each operator runs in its own pod/process with separate memory - FAR and SNR don't share package-level variables)

**Comment by slintes** [comment]
please add [.DS_Store](https://github.com/medik8s/common/pull/27/changes#diff-e2eb93a61ffd7877ea5c751abcb3a618e8e2e9a2073a27f66d4114fe10819f86) to your global `.gitignore` :)

File: pkg/taints/taints.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+type OutOfServiceTaintInfo struct {
+	Supported bool // true if k8s version >= 1.26
+	GA        bool // true if k8s version >= 1.28
+}
+
+var (
+	loggerTaint   = ctrl.Log.WithName("taints")
+	leadingDigits = regexp.MustCompile(`^(\d+)`)
+
+	OutOfServiceInfo OutOfServiceTaintInfo
```
Comment (slintes) [inline pkg/taints/taints.go]: Using a global public var for this is bad design. It can be read and modified by everyone.

File: pkg/taints/taints.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+		if taint.MatchTaint(taintToFind) {
+			return true
+		}
+	}
+	return false
+}
+
+// FilterOutTaint returns a new taint slice without taints matching the given taintToDelete by key and effect.
+// Also returns true if any taints were filtered out, false otherwise.
+func FilterOutTaint(taints []corev1.Taint, taintToDelete *corev1.Taint) ([]corev1.Taint, bool) {
```
Comment (slintes) [inline pkg/taints/taints.go]: `Out` isn't needed, `Filter`  is good enough and best practice. And since the package name is `taints` already, you can even skip `Taint` IMHO. So it will be used as `taints.Filter(...)`
Also: when I see `taintToDelete` (should be `taintToFilter` btw, or just `taint` because its meaning is obvious) as pointer, I would expect that exactly this taint is filtered from the slice. But it's every taint with the same key and effect.
What about this: `Filter(taints []corev1.Taint, key, effect string)`? :thinking:  With that I know what it does without the need to read the doc.

File: pkg/taints/taints.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	var newTaints []corev1.Taint
+	deleted := false
+	for _, taint := range taints {
```
Comment (slintes) [inline pkg/taints/taints.go]: this can be simplified by using the `slices` package

File: pkg/taints/taints.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+var (
+	loggerTaint   = ctrl.Log.WithName("taints")
+	leadingDigits = regexp.MustCompile(`^(\d+)`)
+
+	OutOfServiceInfo OutOfServiceTaintInfo
+)
+
+// TaintExists checks if the given taint exists in list of taints. Returns true if exists false otherwise.
+func TaintExists(taints []corev1.Taint, taintToFind *corev1.Taint) bool {
```
Comment (slintes) [inline pkg/taints/taints.go]: What about `func Contains(taints, key, effect)`?

File: pkg/taints/taints.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+// AppendTaintToNode adds the given taint to the node and patches it.
+// Returns true if the taint was added, false if it already existed.
+// Sets TimeAdded to the current time when adding the taint.
+func AppendTaintToNode(ctx context.Context, c client.Client, node *corev1.Node, taint corev1.Taint) (bool, error) {
+	if TaintExists(node.Spec.Taints, &taint) {
+		return false, nil
+	}
+
+	patch := client.MergeFrom(node.DeepCopy())
```
Comment (slintes) [inline pkg/taints/taints.go]: applying a `MergeFrom` patch can potentially overwrite changes made by other controllers.  Have a look at `StrategicMergeFrom`

File: pkg/taints/taints.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+		Key:    corev1.TaintNodeOutOfService,
+		Value:  "nodeshutdown",
+		Effect: corev1.TaintEffectNoExecute,
+	}
+}
+
+// AppendTaintToNode adds the given taint to the node and patches it.
+// Returns true if the taint was added, false if it already existed.
+// Sets TimeAdded to the current time when adding the taint.
+func AppendTaintToNode(ctx context.Context, c client.Client, node *corev1.Node, taint corev1.Taint) (bool, error) {
```
Comment (slintes) [inline pkg/taints/taints.go]: nit: `Add` would be slightly better, because we don't care about order. `Append` implies adding at the end only.

**Review by slintes — CHANGES_REQUESTED** [review]
I know this probably is a lot of copy paste of existing code. But this is the chance to improve things :) 
Main issue is the use a Init(...) function and storing the result in a global public var. Please rethink this. 
Some more comments inline.

File: pkg/taints/taints.go (hunk 2 above)
Comment (Tzif-Morgen) [reply pkg/taints/taints.go]: Updated to use slices.DeleteFunc and renamed to Filter. 
I kept the Taint object as a parameter since the purpose is to remove a specific taint entity. This allows us to use MatchTaint for the identity check (Key and Effect), which is the K8s logic for matching taints. what do u think?

File: pkg/taints/taints.go (hunk 4 above)
Comment (Tzif-Morgen) [reply pkg/taints/taints.go]: I prefer keeping the Taint object here as well, for the same reason I mentioned above. Since we are checking for the existence of a specific taint entity, passing the object feels more natural IMHO.
Also, by passing the object, we can continue using MatchTaint internally to maintain a single logic for identity checks (Key+Effect) across both Filter and TaintExists (and avoids manually comparing strings in multiple places). I’d love to hear your thoughts on this

File: pkg/taints/taints.go (hunk 2 above)
Comment (slintes) [reply pkg/taints/taints.go]: > This allows us to use MatchTaint for the identity check (Key and Effect), which is the K8s logic for matching taints

ah ok, I missed that this is a k8s function, and it needs the pointer. Ok, wfm 👍🏼

File: pkg/taints/taints.go (hunk 1 above)
Comment (Tzif-Morgen) [reply pkg/taints/taints.go]: changed. 
please also take a look at https://github.com/medik8s/fence-agents-remediation/pull/200 for usage example

File: pkg/taints/taints.go (hunk 4 above)
Comment (slintes) [reply pkg/taints/taints.go]: wdyt about the renaming part, Contains vs TaintExists?

Comment (slintes) [reply pkg/taints/taints.go]: in the slices package the is something for contains IIUC

File: pkg/taints/taints.go (hunk 1 above)
Comment (slintes) [reply pkg/taints/taints.go]: much better 👍🏼

File: pkg/taints/taints.go (hunk 4 above)
Comment (Tzif-Morgen) [reply pkg/taints/taints.go]: changed. wdyt?

Comment (slintes) [reply pkg/taints/taints.go]: 👍🏼

**Review by slintes — DISMISSED** [review]
looks good, will add a hold in case someone else want to do a review

/hold

**Comment by slintes** [comment]
@coderabbitai review

File: pkg/taints/taints.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+	}
+	return true, nil
+}
+
+// DetectOutOfServiceTaintInfoWithRetry detects out-of-service taint support based on k8s version, in case it fails (potentially due to network issues) it will retry for a limited time period.
+func DetectOutOfServiceTaintInfoWithRetry(ctx context.Context, config *rest.Config) (OutOfServiceTaintInfo, error) {
+	var info OutOfServiceTaintInfo
+	var err error
+	interval := 2 * time.Second // retry every 2 seconds
+	timeout := 10 * time.Second // for a period of 10 seconds
```
Comment (clobrano) [inline pkg/taints/taints.go:114]: I would appreciate if the operators could decide the timing independently.
We could extract these values as package-level constants, or (better) pass them as arguments

File: pkg/taints/taints_test.go [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+
+		When("multiple matching taints exist", func() {
+			It("should remove all matching taints", func() {
+				duplicate := corev1.Taint{Key: "key2", Effect: corev1.TaintEffectNoExecute, Value: "different"}
+				taints := []corev1.Taint{taint1, taint2, duplicate, taint3}
+				newTaints, removed := FilterOut(taints, &taint)
+				Expect(removed).To(BeTrue())
+				Expect(newTaints).To(HaveLen(2))
+				Expect(newTaints).To(ContainElement(taint1))
+				Expect(newTaints).To(ContainElement(taint3))
```
Comment (clobrano) [inline pkg/taints/taints_test.go]: Here we should also test that `taint2` and `duplicate` have been removed

File: pkg/taints/taints_test.go [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+			taint := CreateOutOfServiceTaint()
+			Expect(taint.Key).To(Equal(corev1.TaintNodeOutOfService))
+			Expect(taint.Value).To(Equal("nodeshutdown"))
+			Expect(taint.Effect).To(Equal(corev1.TaintEffectNoExecute))
+		})
+	})
+
+	Context("getOutOfServiceTaintInfo", func() {
+		When("version is supported but not GA", func() {
+			DescribeTable("should return Supported=true, GA=false",
```
Comment (clobrano) [inline pkg/taints/taints_test.go:100]: nice :+1:

**Review by clobrano — CHANGES_REQUESTED** [review]
I left a couple of comments. Moreover, also CodeRabbit gave interesting analysis, we should address its remarks :)

File: pkg/taints/taints_test.go (hunk 8 above)
Comment (Tzif-Morgen) [reply pkg/taints/taints_test.go]: Used ConsistOf instead. wdyt?

File: pkg/taints/taints.go [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+	// Using wait.PollUntilContextTimeout to retry detection in case there is a temporary network issue.
+	pollErr := wait.PollUntilContextTimeout(ctx, interval, timeout, true, func(ctx context.Context) (bool, error) {
+		if info, err = detectOutOfServiceTaintInfo(config); err != nil {
+			return false, nil // Keep retrying
+		}
+		return true, nil // Success
+	})
+
+	// Respect context cancellation - return poll error so caller knows operation was cancelled
+	if pollErr != nil && (errors.Is(pollErr, context.Canceled) || errors.Is(pollErr, context.DeadlineExceeded)) {
```
Comment (mshitrit) [inline pkg/taints/taints.go:125]: IIUC this might be a bug.
I'll break down how I understand it and correct me if I'm wrong.

**Desired action:** 
1. In case the context was cancelled outside of this method (likely before the 10 sec window) than we want to return the cancelled context error as it's the most relevant one. 
2.  in case we catch an error during our 10 sec retry window we want to return that (last one caught) error.
(For example if we kept retrying because we got an error X from the API Server than we should return error X)

**Actual action:** 
1. Working as designed
2.  We store the  error we want to return in `err` var, but we never return it because we will exit here - our pollErr would be  of `context.DeadlineExceeded` because our own 10 sec deadline context exceeded.

File: pkg/taints/taints_test.go (hunk 8 above)
Comment (clobrano) [reply pkg/taints/taints_test.go]: Even better

**Comment by Tzif-Morgen** [comment]
Not relevat right now; could be useful later
